pkg/sql/jobs/registry.go

```diff
@@ -54,3 +94,202 @@ func (r *Registry) LoadJob(ctx context.Context, jobID int64) (*Job, error) {
+var DefaultAdoptInterval = 30 * time.Second
+
+// Start polls the current node for liveness failures and cancels all registered
+// jobs if it observes a failure.
+func (r *Registry) Start(
+	stopper *stop.Stopper, nl nodeLiveness, cancelInterval, adoptInterval time.Duration,
+) error {
+	stopper.RunWorker(context.Background(), func(ctx context.Context) {
+		for {
+			select {
+			case <-time.After(cancelInterval):
+				r.maybeCancelJobs(ctx, nl)
+			case <-stopper.ShouldStop():
+				return
+			}
+		}
+	})
+
+	stopper.RunWorker(context.Background(), func(ctx context.Context) {
+		for {
+			select {
+			case <-time.After(adoptInterval):
+				if err := r.maybeAdoptJob(ctx, nl); err != nil {
+					log.Errorf(ctx, "error while adopting jobs: %+v", err)
+				}
+			case <-stopper.ShouldStop():
+				return
+			}
+		}
+	})
+	return nil
+}
+
+func (r *Registry) maybeCancelJobs(ctx context.Context, nl nodeLiveness) {
+	liveness, err := nl.Self()
+	if err != nil {
+		log.Warningf(ctx, "unable to get node liveness: %s", err)
+		// Conservatively assume our lease has expired. Abort all jobs.
+		r.mu.Lock()
+		defer r.mu.Unlock()
+		r.cancelAll(ctx)
+		return
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	// TODO(benesch): cancel only jobs from previous epochs. This logic is correct
+	// but too aggressive, as it might cancel jobs created immediately after a
+	// liveness failure.
+	sameEpoch := liveness.Epoch == r.mu.epoch
+	if !sameEpoch || !liveness.IsLive(r.clock.Now(), r.clock.MaxOffset()) {
+		r.cancelAll(ctx)
+		r.mu.epoch = liveness.Epoch
+	}
+}
+
+type resumeHookFn func(Type) func(context.Context, *Job) error
+
+var resumeHooks []resumeHookFn
+
+// AddResumeHook adds a resume hook.
+func AddResumeHook(fn resumeHookFn) {
+	resumeHooks = append(resumeHooks, fn)
+}
+
+func (r *Registry) maybeAdoptJob(ctx context.Context, nl nodeLiveness) error {
+	var rows []parser.Datums
+	if err := r.db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
+		var err error
+		const stmt = `SELECT id, payload FROM system.jobs WHERE status IN ($1, $2) ORDER BY created DESC`
+		rows, err = r.ex.QueryRowsInTransaction(ctx, "adopt-job", txn, stmt, StatusPending, StatusRunning)
+		return err
+	}); err != nil {
+		return err
+	}
+
+	type nodeStatus struct {
+		isLive bool
+		epoch  int64
+	}
+	nodeStatusMap := make(map[roachpb.NodeID]*nodeStatus)
+	{
+		now, maxOffset := r.clock.Now(), r.clock.MaxOffset()
+		for _, liveness := range nl.GetLivenesses() {
+			nodeStatusMap[liveness.NodeID] = &nodeStatus{
+				isLive: liveness.IsLive(now, maxOffset),
+				epoch:  liveness.Epoch,
+			}
+		}
+	}
+
+	for _, row := range rows {
+		id := (*int64)(row[0].(*parser.DInt))
+		payload, err := UnmarshalPayload(row[1])
+		if err != nil {
+			return err
+		}
+
+		if log.V(2) {
+			log.Infof(ctx, "evaluating job %d", *id)
+		}
+
+		if payload.Lease == nil {
+			// If the lease is missing, it simply means the job does not yet support
```

Comment:
Should you do anything here to "cancel" jobs that are in the running state but won't ever finish? Or is there some other way for old abandoned jobs to get cleaned up?